langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Sort.java

```diff
@@ -18,73 +18,185 @@
 
 package org.ballerinalang.langlib.array;
 
+import org.ballerinalang.jvm.BallerinaErrors;
 import org.ballerinalang.jvm.scheduling.Strand;
 import org.ballerinalang.jvm.types.BArrayType;
+import org.ballerinalang.jvm.types.BFunctionType;
+import org.ballerinalang.jvm.types.BType;
+import org.ballerinalang.jvm.types.BUnionType;
+import org.ballerinalang.jvm.types.TypeTags;
 import org.ballerinalang.jvm.values.ArrayValue;
-import org.ballerinalang.jvm.values.ArrayValueImpl;
 import org.ballerinalang.jvm.values.FPValue;
 import org.ballerinalang.model.types.TypeKind;
 import org.ballerinalang.natives.annotations.Argument;
 import org.ballerinalang.natives.annotations.BallerinaFunction;
 import org.ballerinalang.natives.annotations.ReturnType;
 
+import java.math.BigDecimal;
+import java.util.Optional;
+
 import static org.ballerinalang.jvm.values.utils.ArrayUtils.checkIsArrayOnlyOperation;
 import static org.ballerinalang.util.BLangCompilerConstants.ARRAY_VERSION;
 
 /**
- * Native implementation of lang.array:sort((any|error)[], function).
+ * Native implementation of lang.array:sort((any|error)[], direction, function).
  *
  * @since 1.0
  */
 @BallerinaFunction(
         orgName = "ballerina", packageName = "lang.array", version = ARRAY_VERSION, functionName = "sort",
-        args = {@Argument(name = "arr", type = TypeKind.ARRAY), @Argument(name = "func", type = TypeKind.FUNCTION)},
+        args = {@Argument(name = "arr", type = TypeKind.ARRAY), @Argument(name = "direction", type = TypeKind.OBJECT),
+                @Argument(name = "func", type = TypeKind.OBJECT)},
         returnType = {@ReturnType(type = TypeKind.ARRAY)},
         isPublic = true
 )
 public class Sort {
 
-    public static ArrayValue sort(Strand strand, ArrayValue arr, FPValue<Object, Long> func) {
+    public static ArrayValue sort(Strand strand, ArrayValue arr, Object direction, Object func) {
         checkIsArrayOnlyOperation(arr.getType(), "sort()");
-        ArrayValue aux = new ArrayValueImpl((BArrayType) arr.getType());
-        mergesort(arr, aux, 0, arr.size() - 1, strand, func);
+        FPValue<Object, Object> function = (FPValue<Object, Object>) func;
+        BType elemType = ((BArrayType) arr.getType()).getElementType();
+        boolean isAscending = true;
+        if (direction.toString().equals("descending")) {
+            isAscending = false;
+        }
+
+        Object[][] sortArr = new Object[arr.size()][2];
+        Object[][] sortArrClone = new Object[arr.size()][2];
+        for (int i = 0; i < arr.size(); i++) {
+            if (function != null) {
+                Object x = function.call(new Object[]{strand, arr.get(i), true});
+                sortArr[i][0] = x;
+            } else {
+                sortArr[i][0] = arr.get(i);
+            }
+            sortArr[i][1] = arr.get(i);
+        }
```

Comment:
Since `function` doesn't change for a single invocation, wouldn't it be better to do 

```suggestion
        if (function != null) {
            for (int i = 0; i < arr.size(); i++) {
                sortArr[i][0] = function.call(new Object[]{strand, arr.get(i), true});
                sortArr[i][1] = arr.get(i);
            }
        } else {
            for (int i = 0; i < arr.size(); i++) {
                sortArr[i][0] = sortArr[i][1] = arr.get(i);
            }
        }
```